### PR TITLE
Document partial_order_concurrency [DOC-140]

### PR DESCRIPTION
**Pull request by xbauch**
This PR is only a part of DOC-140. There are no functional changes to the code.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/partial_order_concurrency.h [hunk 1]
```diff
@@ -35,6 +38,10 @@ class partial_order_concurrencyt:public messaget
     AX_PROPAGATION=8
   };
 
+  /// Build identifier for the read/write clock variable
+  /// \param e: either shared read or shared write event
+  /// \param axiom: the clock variable to be used
+  /// \return irep id representing the clock variable of the event
```
Comment (tautschnig) [inline src/goto-symex/partial_order_concurrency.h]: I'd use "Identifier" instead of "irep id"

File: src/goto-symex/partial_order_concurrency.h [hunk 2]
```diff
@@ -53,40 +60,73 @@ class partial_order_concurrencyt:public messaget
...
   void add_init_writes(symex_target_equationt &);
 
   // a per-thread numbering of the events
   typedef std::map<event_it, unsigned> numberingt;
   numberingt numbering;
 
-  // produces the symbol ID for an event
+  /// Produce the symbol ID for an event
+  /// \param event: SSA step for the event
+  /// \return identifier of the lhs
```
Comment (tautschnig) [inline src/goto-symex/partial_order_concurrency.h]: I'd drop "of the lhs"

File: src/goto-symex/partial_order_concurrency.h [hunk 3]
```diff
@@ -53,40 +60,73 @@ class partial_order_concurrencyt:public messaget
...
   static inline irep_idt id(event_it event)
   {
     return event->ssa_lhs.get_identifier();
   }
 
-  // produces an address ID for an event
+  /// Produce an address ID for an event
+  /// \param event: SSA step for the event
+  /// \return identifier of the lhs (without L2)
```
Comment (tautschnig) [inline src/goto-symex/partial_order_concurrency.h]: I'd say "L1-renamed identifier"

**Review by tautschnig — APPROVED** [review]

**Comment by xbauch** [comment]
@tautschnig Thanks for the quick review, I rewrote the documentation as suggested.

File: src/goto-symex/partial_order_concurrency.h [hunk 4]
```diff
@@ -35,6 +38,10 @@ class partial_order_concurrencyt:public messaget
     AX_PROPAGATION=8
   };
 
+  /// Build identifier for the read/write clock variable
+  /// \param e: either shared read or shared write event
+  /// \param axiom: the clock variable to be used
```
Comment (danpoe) [inline src/goto-symex/partial_order_concurrency.h]: I think this parameter indicates the axiom that the resulting clock id is part of. Would that be right @tautschnig?

File: src/goto-symex/README.md [hunk 5]
```diff
@@ -276,3 +276,10 @@ digraph G {
   2 -> 3 [label="counter-examples"];
 }
 \enddot
+
+\section Concurrency
+
+\subsection Partial Order Reduction
```
Comment (danpoe) [inline src/goto-symex/README.md]: I'd name it just `Partial Order Concurrency` as it's not really partial order reduction.

File: src/goto-symex/partial_order_concurrency.h (hunk 4 above)
Comment (tautschnig) [reply src/goto-symex/partial_order_concurrency.h]: Yes.

File: src/goto-symex/README.md (hunk 5 above)
Comment (xbauch) [reply src/goto-symex/README.md]: You mean that its used to implement memory models but the ordering is never used for reduction?

File: src/goto-symex/partial_order_concurrency.h (hunk 4 above)
Comment (xbauch) [reply src/goto-symex/partial_order_concurrency.h]: I agree; should I make it explicit in the description?

File: src/goto-symex/README.md (hunk 5 above)
Comment (peterschrammel) [reply src/goto-symex/README.md]: Cf. http://www.kroening.com/papers/cav2013-wpo.pdf

File: src/goto-symex/partial_order_concurrency.h [hunk 6]
```diff
@@ -53,40 +60,73 @@ class partial_order_concurrencyt:public messaget
...
-  // produce a clock symbol for some event
   typet clock_type;
+
+  /// Produce a clock symbol for some event
+  /// \param e: event is either shared read/write or spawn
+  /// \param axiom: clock variable
+  /// \return symbol of type _clock_type_ with id from \ref rw_clock_id
   symbol_exprt clock(event_it e, axiomt axiom);
+
+  /// Initialise the __clock_type__ so that it can be used to number events
```
Comment (peterschrammel) [inline src/goto-symex/partial_order_concurrency.h]: Initialize

**Review by peterschrammel — APPROVED** [review]

File: src/goto-symex/README.md (hunk 5 above)
Comment (danpoe) [reply src/goto-symex/README.md]: It's a different technique. The encoding starts with unconstrained reads (i.e., with shared reads essentially being nondet), and then additional constraints are added to make the model stronger to yield e.g. TSO or SC.

File: src/goto-symex/partial_order_concurrency.h (hunk 4 above)
Comment (danpoe) [reply src/goto-symex/partial_order_concurrency.h]: Yes, I think we should make it explicit for all the appearances of `axiom`.

File: src/goto-symex/partial_order_concurrency.h (hunk 6 above)
Comment (xbauch) [reply src/goto-symex/partial_order_concurrency.h]: Well I guess it is called Oxford spelling after all. Spelling changed.

**Comment by xbauch** [comment]
I see, so the partial ordering is used to replace thread interleaving rather than reduce it. I changed the sub-section title in the readme.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0f4d498cfef2d189d79884ee5b495d9ec8ccc101).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99823236

**Pull request merged**